sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala

```diff
@@ -179,12 +189,16 @@ object SchemaPruning extends Rule[LogicalPlan] {
     // We need to replace the expression ids of the pruned relation output attributes
     // with the expression ids of the original relation output attributes so that
     // references to the original relation's output are not broken
-    val outputIdMap = output.map(att => (att.name, att.exprId)).toMap
+
+    // We also need to keep the metadata of the attribute,
+    // so that references to the metadata struct is not broken
+    val nameAttributeMap = output.map(att => (att.name, att)).toMap
     requiredSchema
       .toAttributes
       .map {
-        case att if outputIdMap.contains(att.name) =>
-          att.withExprId(outputIdMap(att.name))
+        case att if nameAttributeMap.contains(att.name) =>
+          val outputAttribute = nameAttributeMap(att.name)
+          att.withExprId(outputAttribute.exprId).withMetadata(outputAttribute.metadata)
```

Comment:
Since we are touching this, I think a better way to fix this is to only pick the data type from the pruned schema:
```
outputAttribute.withDataType(attr.dataType)
```
This makes sure that the only change here is to prune the schema and update the data type of attributes.

Reply:
make sens! done!